### PR TITLE
More realistic sort benchmarks

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Relates to #5879

# Rationale for this change

This combines the merge and sort benchmarks together to avoid code duplication, it also makes a number of changes to make the benchmarks more realistic:

* Each partition / stream now contains multiple batches, an average of 13
* The sort_merge bench now performs a partitioned sort instead of sorting into a single partition and then running a SortPreservingMerge on the single partition (which is a no-op)
* Tuples are sliced and then collected into their output arrays, this ensures that DictionaryArray aren't all using the same identical dictionary values
* Merge benchmark is run on presorted data, with the other benchmarks now run on unsorted data

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/benches/sort.rs [hunk 1]
```diff
@@ -41,167 +94,80 @@ use rand::rngs::StdRng;
...
-
-// * physical sort expr (X, Y Z, NULLS FIRST, ASC) (not parameterized)
-//
-// streams of distinct values
-// streams with 10% duplicated values (within each stream, and across streams)
-// These cases are intended to model important usecases in TPCH
-// parameters:
-//
-// Input schemas
-lazy_static! {
```
Comment (tustvold) [inline datafusion/core/benches/sort.rs:62]: Lazy static seemed overkill for what this was doing, I opted to keep things simple

File: datafusion/core/benches/sort.rs [hunk 2]
```diff
@@ -210,58 +178,61 @@ impl SortBenchCase {
...
-/// Encapsulates running a test case where input partitioning is not preserved.
-struct SortBenchCasePreservePartitioning {
-    runtime: Runtime,
-    task_ctx: Arc<TaskContext>,
 
-    // The plan to run
-    plan: Arc<dyn ExecutionPlan>,
-    partition_count: usize,
-}
+    fn sort(partitions: &[Vec<RecordBatch>]) -> Self {
```
Comment (alamb) [inline datafusion/core/benches/sort.rs:201]:
```suggestion
    /// Test SortExec in "non partitioned" mode which sorts the input streams
    /// into a single sorted output stream
    fn sort(partitions: &[Vec<RecordBatch>]) -> Self {
```

File: datafusion/core/benches/sort.rs [hunk 3]
```diff
@@ -210,58 +178,61 @@ impl SortBenchCase {
...
+        let plan = Arc::new(SortExec::try_new(sort, exec, None).unwrap());
+
+        Self {
+            runtime,
+            task_ctx,
+            plan,
+        }
+    }
+
+    fn sort_partitioned(partitions: &[Vec<RecordBatch>]) -> Self {
```
Comment (alamb) [inline datafusion/core/benches/sort.rs:220]:
```suggestion
    /// Test SortExec in "partitioned" mode which sorts the input streams
    /// individually into some number of output streams
    fn sort_partitioned(partitions: &[Vec<RecordBatch>]) -> Self {
```

File: datafusion/core/benches/sort.rs [hunk 4]
```diff
@@ -210,58 +178,61 @@ impl SortBenchCase {
         }
     }
 
-    /// runs the specified plan to completion, draining all input and
-    /// panic'ing on error
-    fn run(&self) {
-        let plan = Arc::clone(&self.plan);
-        let task_ctx = Arc::clone(&self.task_ctx);
+    fn sort_merge(partitions: &[Vec<RecordBatch>]) -> Self {
```
Comment (alamb) [inline datafusion/core/benches/sort.rs:181]:
```suggestion
    /// Test SortExec in  "partitioned" mode followed by a SortPreservingMerge
    fn sort_merge(partitions: &[Vec<RecordBatch>]) -> Self {
```

File: datafusion/core/benches/sort.rs [hunk 5]
```diff
@@ -210,58 +178,61 @@ impl SortBenchCase {
...
-        self.runtime.block_on(async move {
-            let mut stream = plan.execute(0, task_ctx).unwrap();
-            while let Some(b) = stream.next().await {
-                b.expect("unexpected execution error");
-            }
-        })
+        let exec = MemoryExec::try_new(partitions, schema, None).unwrap();
+        let exec =
+            SortExec::new_with_partitioning(sort.clone(), Arc::new(exec), true, None);
+        let plan = Arc::new(SortPreservingMergeExec::new(sort, Arc::new(exec)));
```
Comment (alamb) [inline datafusion/core/benches/sort.rs:192]: I would expect this to behave the same, performance wise, as `sort` -- is that your expectation too?

File: datafusion/core/benches/sort.rs [hunk 6]
```diff
@@ -41,167 +94,82 @@ use rand::rngs::StdRng;
...
     task_ctx: Arc<TaskContext>,
 
     // The plan to run
     plan: Arc<dyn ExecutionPlan>,
 }
 
-impl SortBenchCase {
+impl BenchCase {
     /// Prepare to run a benchmark that merges the specified
     /// partitions (streams) together using all keyes
```
Comment (alamb) [inline datafusion/core/benches/sort.rs:162]:
```suggestion
    /// pre-sorted partitions (streams) together using all keys
```

**Review by alamb — APPROVED** [review]
Thank you @tustvold  -- I reviewed this and it makes sense to me. I also spot checked the inputs (by printing them out to standard out and it looked good)

cc @jaylmiller

File: datafusion/core/benches/sort.rs (hunk 5 above)
Comment (tustvold) [reply datafusion/core/benches/sort.rs:192]: Eventually yes, currently for single column cases it performs worse

**Comment by jaylmiller** [comment]
> Thank you @tustvold -- I reviewed this and it makes sense to me. I also spot checked the inputs (by printing them out to standard out and it looked good)
> 
> cc @jaylmiller

Looks good to me as well!

**Pull request merged**